scripts/dev/tidy.php

```diff
@@ -0,0 +1,310 @@
+    }
+
+    /**
+     * Log error information to STDOUT.
+     */
+    public function error(string $file, string $message): string
+    {
+        return $file.': '.$this->red.$message.$this->default."\n";
+    }
+}
+
+// Process arguments
+if ($argc < 2) {
+    echo usage();
+    exit;
+}
+
+array_shift($argv);
+
+foreach($argv as $argument) {
+    switch($argument) {
+        case '-t':
+        case '--trim':
+            $trim = true;
+        break;
+        case '--no-whitespace':
+            $whitespace = false;
+        break;
+        case '--no-eol':
+            $eol = false;
+        break;
+        case '-h':
+        case '--help':
+            echo usage();
+            exit;
+        break;
+        case (preg_match('/^-.+/', $argument) ? true : false):
+        break;
+        case (file_exists($argument)):
+            $path = $argument;
+        break;
+        default:
+            echo usage();
+            exit;
+        break;
+    }
+}
+
+$trim = $trim ?? false;
+$whitespace = $whitespace ?? true;
+$eol = $eol ?? true;
+if (!isset($path)) die("Error: missing path");
+
+$filter = function ($file, $key, $iterator) use ($excludePaths, $excludePatterns, $path) {
+    $pathnameWithoutBase = str_replace($path.'/', '', $file->getPathname());
+
+    foreach ($excludePatterns as $regex) {
+        if (preg_match($regex, $pathnameWithoutBase)) {
+            return false;
+        }
+    }
+
+    return (!in_array($pathnameWithoutBase, $excludePaths));
+};
+
+/**
+ * Script execution
+ */
+
+$files = [];
+
+if (is_file($path)) {
+    $files = [$path];
+} else {
+    $innerIterator = new \RecursiveDirectoryIterator(
+        $path,
+        \RecursiveDirectoryIterator::SKIP_DOTS
+    );
+
+    $iterator = new \RecursiveIteratorIterator(
+        new \RecursiveCallbackFilterIterator($innerIterator, $filter)
+    );
+
+    foreach ($iterator as $pathName => $fileInfo) {
+        $files[] = $pathName;
+    }
+}
+
+$fixer = new Fixer();
+$logger = new Logger();
+
+echo "Checking for file(s) needing whitespace trimmed and final new lines fixed\n";
+
+foreach ($files as $file) {
+    $content = file_get_contents($file);
+
+    $trimmed = ($whitespace) ? $fixer->trimWhitespace($content) : $content;
+
+    // Whitespace
+    if ($whitespace && !$trim && $trimmed !== $content) {
+        echo $logger->error($file, 'includes whitespace');
+    } elseif ($whitespace && $trimmed !== $content) {
+        echo $logger->info($file, 'whitespace trimmed');
+    }
+
+    // Final new lines
+    if ($eol && !$trim && $fixer->trimFinalNewLine($content) !== $content) {
+        echo $logger->error($file, 'needs final EOL fixed');
+    } elseif ($eol && $fixer->trimFinalNewLine($content) !== $content) {
+        $trimmed = $fixer->trimFinalNewLine($trimmed);
+        echo $logger->info($file, 'final EOL fixed');
+    }
+
+    // Overwrite file if trim option has been specified
+    if($trim && file_put_contents($file, $trimmed) === false) {
```

Comment:
I'd suggest checking against the original file here first, to avoid unnecessary timestamp updates, which confuse VCS, incremental compilation etc.

Reply:
Fixed. Thanks for the review...